From 89d2b52bed55465d1c8a29221db6f802d224794b Mon Sep 17 00:00:00 2001
From: Omair Kamil <okamil@google.com>
Date: Fri, 4 Oct 2024 12:17:54 -0700
Subject: [PATCH 1/9] Interop fix to not read PPCP for devices that report
 incompatible values.

Bug: 360314855
Fix: 360314855
Flag: EXEMPT, targeted IOP fix
Test: m Bluetooth, manual testing
(cherry picked from https://android-review.googlesource.com/q/commit:99527fc62ec1f891cd53afe909bc48c5a3d710bb)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:16f1d33299aa6918faea4720604094eea8851c8f)
Merged-In: I69b13f959351075b45603a514f06a967101cccf4
Change-Id: I69b13f959351075b45603a514f06a967101cccf4
---
 system/bta/dm/bta_dm_disc.cc       | 13 ++++++++++++-
 system/conf/interop_database.conf  |  3 +++
 system/device/test/interop_test.cc |  3 +++
 3 files changed, 18 insertions(+), 1 deletion(-)

diff --git a/system/bta/dm/bta_dm_disc.cc b/system/bta/dm/bta_dm_disc.cc
index b8e24551db..3e92f3e978 100644
--- a/system/bta/dm/bta_dm_disc.cc
+++ b/system/bta/dm/bta_dm_disc.cc
@@ -33,9 +33,11 @@
 #include "bta/dm/bta_dm_disc_legacy.h"
 #include "bta/include/bta_gatt_api.h"
 #include "com_android_bluetooth_flags.h"
+#include "btif/include/btif_storage.h"
 #include "common/circular_buffer.h"
 #include "common/init_flags.h"
 #include "common/strings.h"
+#include "device/include/interop.h"
 #include "internal_include/bt_target.h"
 #include "main/shim/dumpsys.h"
 #include "os/logging/log_adapter.h"
@@ -333,8 +335,17 @@ static void bta_dm_disc_result(tBTA_DM_SVC_RES& disc_result) {
     bta_dm_discovery_cb.service_search_cbacks.on_service_discovery_results(
         r.bd_addr, r.uuids, r.result);
   } else {
+    char remote_name[BD_NAME_LEN] = "";
     bta_dm_discovery_cb.transports &= ~BT_TRANSPORT_LE;
-    GAP_BleReadPeerPrefConnParams(bta_dm_discovery_cb.peer_bdaddr);
+    if (btif_storage_get_stored_remote_name(bta_dm_discovery_cb.peer_bdaddr, remote_name) &&
+        interop_match_name(INTEROP_DISABLE_LE_CONN_PREFERRED_PARAMS, remote_name)) {
+      // Some devices provide PPCP values that are incompatible with the device-side firmware.
+      log::info("disable PPCP read: interop matched name {} address {}", remote_name,
+                bta_dm_discovery_cb.peer_bdaddr);
+    } else {
+      log::info("reading PPCP");
+      GAP_BleReadPeerPrefConnParams(bta_dm_discovery_cb.peer_bdaddr);
+    }
 
     bta_dm_discovery_cb.service_search_cbacks.on_gatt_results(
         bta_dm_discovery_cb.peer_bdaddr, BD_NAME{}, disc_result.gatt_uuids,
diff --git a/system/conf/interop_database.conf b/system/conf/interop_database.conf
index a275f6bc6b..8f190ab9ee 100644
--- a/system/conf/interop_database.conf
+++ b/system/conf/interop_database.conf
@@ -194,6 +194,9 @@ Motorola Keyboard KZ500 v122 = Name_Based
 [INTEROP_DISABLE_LE_CONN_PREFERRED_PARAMS]
 BSMBB09DS = Name_Based
 ELECOM = Name_Based
+Dexcom = Name_Based
+DXCM = Name_Based
+DX0 = Name_Based
 
 # Disable role switch for headsets/car-kits
 # Some car kits allow role switch but when DUT initiates role switch
diff --git a/system/device/test/interop_test.cc b/system/device/test/interop_test.cc
index 3ad3db2ce9..24c2c2ab37 100644
--- a/system/device/test/interop_test.cc
+++ b/system/device/test/interop_test.cc
@@ -433,6 +433,9 @@ TEST_F(InteropTest, test_name_hit) {
                                  "Motorola Keyboard KZ500"));
   EXPECT_TRUE(interop_match_name(INTEROP_DISABLE_LE_CONN_PREFERRED_PARAMS,
                                  "BSMBB09DS"));
+  EXPECT_TRUE(interop_match_name(INTEROP_DISABLE_LE_CONN_PREFERRED_PARAMS, "DXCMog"));
+  EXPECT_TRUE(interop_match_name(INTEROP_DISABLE_LE_CONN_PREFERRED_PARAMS, "Dexcom 123"));
+  EXPECT_TRUE(interop_match_name(INTEROP_DISABLE_LE_CONN_PREFERRED_PARAMS, "DX01ab"));
   EXPECT_TRUE(interop_match_name(INTEROP_DISABLE_AAC_CODEC, "abramtek M1"));
   EXPECT_TRUE(
       interop_match_name(INTEROP_DISABLE_AAC_VBR_CODEC, "Audi_MMI_2781"));
-- 
2.34.1

